Group and eperson management

benbosman opened the pull request:
Related ticket: https://jira.lyrasis.org/browse/DS-4026
* Created a Rest Contract for EPerson Groups (some of the API behavior already exists in the API)
* Updated the Rest Contract for EPersons (some of the API behavior already exists in the API)
* Created a Rest Contract for adding child Groups to parent groups (and deleting, querying)
* Created a Rest Contract for adding EPersons to parent groups (and deleting, querying)

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,70 @@
+      "href": "https://dspace7-internal.atmire.com/rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc/groups"
+    },
+    "self": {
+      "href": "https://dspace7-internal.atmire.com/rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc"
```
Please use the official REST & UI demo web sites for examples. This will allow us to easily replace the reference to dspace.org if needed

benbosman replied:
This was solved

abollini commented on epersongroups.md:
````diff
@@ -0,0 +1,70 @@
+
+```json
+{
+  "name": "Administrator",
````
I suggest to remove the name attribute from the example as it is a "read-only" attribute derived from the dc.title

benbosman replied:
This was solved

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,70 @@
+  "metadata": [
+    {
+      "key": "dc.title",
+      "value": "Administrator"
```
I suggest to use a "My Group" or other generic name instead one close to the special "Administrators"

benbosman replied:
This was solved

abollini commented on epersons.md:
```diff
@@ -2,10 +2,54 @@
+  "type": "eperson",
+  "_links": {
+    "self": {
+      "href": "https://dspace7-internal.atmire.com/rest/api/eperson/epersons/028dcbb8-0da2-4122-a0ea-254be49ca107"
```
can you supply an example from the official REST demo? what about https://dspace7.4science.it/dspace-spring-rest/#https://dspace7.4science.it/dspace-spring-rest/api/eperson/epersons/15508d07-d9ee-4c42-acbd-72fe9a013504

benbosman replied:
This was solved

abollini commented on group2eperson.md:
```diff
@@ -0,0 +1,98 @@
+# Group 2 EPerson mapping Endpoints
```
we don't need a new endpoint for that. We should use an association endpoint inside the groups and epersons endpoints, we can eventually decide to implement the mapping only from one side (in such case my preference goes to the groups endpoint as here we will have in any case the "subgroups").

For example to add a subgroup to a group a text-uri/list POST request should be sent to `/api/eperson/groups/<:uuid>/groups`

to add an eperson we should introduce a new link epersons in the groups endpoint and sent a POST request to `/api/eperson/groups/<:uuid>/epersons`
with the request body to be the URI of the subgroup or eperson (i.e. `https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/<:subgroup-uuid>` or `https://dspace7.4science.it/dspace-spring-rest/api/eperson/epersons/<:eperson-uuid>`

A PUT request can be used to fully override the mapping if needed. See #44 for an example

benbosman replied:
This was solved

AlexanderS commented on epersongroups.md:
```diff
@@ -0,0 +1,70 @@
+        "totalPages": 1,
+        "totalElements": 0
+      }
+    }
```
I would expect something like `epersons` or `members` here.

benbosman replied:
The HAL links to `epersons` and `groups` have been included

AlexanderS commented on epersons.md:
```diff
@@ -2,10 +2,54 @@
+  "email": "user@institution.edu",
+  "requireCertificate": false,
+  "selfRegistered": true,
+  "groups": null,
```
Is this a list of group names or a href-list?

benbosman replied:
This was replaced with a HAL link

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,310 @@
+[Back to the list of all defined endpoints](endpoints.md)
+
+## Main Endpoint
+**GET /api/eperson/groups**
```
We have an eperson category used for epersons and groups. The endpoint stated by @benbosman is what we have currently implemented, the endpoints.md is wrong.
@benbosman can you please fix it as part of this PR?

benbosman replied:
This is already the current endpoint as documented in https://github.com/DSpace/Rest7Contract/blob/master/epersons.md and used in the implementation.

@tdonohue do you agree it's https://github.com/DSpace/Rest7Contract/blob/master/endpoints.md which should be updated?
Also, is there a guideline on which category to use, I didn't see it in the readme

tdonohue replied:
I had initially overlooked that this endpoint is already `/api/eperson/groups`. 

As `EPerson` is also a [type of DSpaceObject](https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/eperson/EPerson.java#L46), I don't understand the purpose to creating a separate "eperson" category here for EPerson and Groups. It results in very odd looking request URIs, in my opinion.

If we have a clear reason, then let me know.  Otherwise, my assumption has been that all DSpace objects appear as `/api/core/{model}/` paths, which would imply this should be `/api/core/epersons/`.  That said, this decision isn't a fault of this PR, so we can do that refactor in a separate PR as needed.

The README doesn't have documented categories yet beyond what is in https://github.com/DSpace/Rest7Contract/blob/master/endpoints.md

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,310 @@
+**GET /api/eperson/groups/<:uuid>/groups**
+
+Filter using parameter:
+* all : retrieve indirect groups as well
```
I share the same doubt than @tdonohue 
The "all" filter should be replaced with a search method or a "projection" once implemented to declare with depth of embedding you want in the response

benbosman replied:
When comparing it with `/api/core/communities/search/subCommunities?parent=<:uuid>`, doesn't that imply `/api/eperson/groups/<:uuid>/groups` without the parameter would also become a search?

tdonohue replied:
@benbosman : I'm less concerned with the `/api/eperson/groups/<:uuid>/groups` as there's a clear "membership" between the parent group and the sub group.  However, if you wanted to refactor that into a `/search` endpoint, I'm OK with that.  It'd likely end up being `/api/eperson/groups/search/subGroups?parent=<:uuid>` then.

My question about the purpose of `?all=true` still exists though. I don't understand what that param is doing and how it is expected to be used in the Angular UI.  So, it's difficult for me to recommend exactly how we might rethink this endpoint.

benbosman replied:
@tdonohue It seems I lost the primary use case for the `?all=true` in the refactoring at https://github.com/DSpace/Rest7Contract/pull/41/commits/c614a94ecfc983862a02ff3fa79e7690e4baadac
The previous REST Contract contained an option to get all the groups of an EPerson, e.g. similar to what is included on the profile page. Having the ability to see all groups an EPerson belongs to is frequently useful. I'll need to recreate the Contract to get the groups of an EPerson.
I agree the `?all=true` parameter would indeed make more sense as a search.

I would also like to know whether you see a substantial difference between a community and sub communities (implemented as a search) vs a group and sub groups (currently in the contract without a search). I just want to know whether there's a particular difference between both cases requiring them to be treated differently, or whether your opinion is that in both cases, both approaches would be good solutions and there's no need for them to be consistent.

tdonohue replied:
Hi @benbosman : since we are already talking about the need to have a `/search` method for the `?all=true` use case, it seems reasonable to also implement the subgroup endpoint as `/api/eperson/groups/search/subGroups?parent=<:uuid>`. 

As you noted, it also keeps this new endpoint consistent with the subcommunities endpoint.

To be honest, I had not noticed this inconsistency between subgroups and subcommunities until you pointed it out.  But, I agree with you that it's probably best to align them.

abollini replied:
we should have both. The main purpose of the /api/eperson/groups/:uuid/groups endpoint is to manage the association. We need to support DELETE, POST and optionally PUT to this endpoint to give the ability to add groups as subgroups.

Using the LinkRest annotation with a linkMethod we should be able to avoid code duplication and just use the same method for the search and for the link, see https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java#L717

benbosman replied:
Implementing both a search and a direct get to retrieve subgroups doesn't make any sense either. The data should be retrievable using a single endpoint.

We should either:
* Remain consistent with the current subcommunities functionality
  * This would use `/api/eperson/groups/search/subGroups?parent=<:uuid>` as suggested by @tdonohue 
  * This doesn't need to impact DELETE and POST on `/api/eperson/groups/:uuid/groups` as long as it's clear this endpoint doesn't support GET (similar to `DELETE /api/core/items/<:uuid>/mappedCollections/<:collection_uuid>` not being supported as a GET in https://github.com/DSpace/Rest7Contract/blob/master/items.md#mapped-collections)
* Update the subcommunities endpoint
  * The https://github.com/DSpace/Rest7Contract/blob/master/communities.md#subcommunities-1 endpoint should be updated to use `/api/core/communities/<:uuid>/subcommunities`. And future potentials DELETE and POST would also happen on `/api/core/communities/<:uuid>/subcommunities`

tdonohue replied:
Looking back on this discussion, I'm having a hard time providing a direction/path forward.

I'd still like to understand the `?all=true` filter better...If that is needed, then I think it's *definitely* a search (and not a filter -- a filter should limit results, and not *expand* the number of results).

So, at a minimum, I think we need a `/api/eperson/groups/search/subGroups` endpoint to support the concept of `all=true`.  However, that does **not** require that we remove all GET support on `/api/eperson/groups/:uuid/groups`, instead we'd just need to document that this endpoint *only* returns subgroups that are *directly members of* the given group.  We can then point folks at the `/api/eperson/groups/search/subGroups` endpoint if they want to perform a different type of search within groups.

While alignment with Communities/SubCommunities is "nice", I don't think it's 100% required here.  Relationships between Communities & SubCommunities are quite different from relationships between Groups. Keep in mind, the Group <-> Group association is many-to-many, while Community <-> Community is *one-to-many* (a subcommunity never belongs to multiple parents, but a Community can have many subcommunities).  So, we shouldn't be treating these concepts as 100% identical... general alignment is nice, but not required.

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,310 @@
+* 204: if the update succeeded 
+* 401 Forbidden - if you are not authenticated
+* 403 Unauthorized - if you are not logged in with sufficient permissions 
+* 422: if the specified group is not found
```
it could be useful to state 404 explicitly to discriminate between the case that the "parent" group is not found and the 422 child group is not found

What if the child cannot be added to the parent due to a cyclic reference (ie. A -> B -> C, -> mean contains we cannot add A to C)? 409 Conflict could be appropriate here. It will be ideal if we report to the client that A should be removed of B to be added to C

benbosman replied:
404 and 409 can indeed be useful
@abollini can you elaborate on how it should be further reported to the client that A should be removed of B to be added to C? Do you have an example for this

abollini replied:
We can start with just reporting that as plain text, it is similar to what we do when an eperson cannot be deleted due to db constraints.
Something like: The group A cannot be added as it is a parent of group C via group B will be ideal but we can also just start with "The group cannot be added due to a cyclic reference"

BTW, looking for an example I found that EPerson delete should be treated in the same way, see for instance
https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/EPersonRestRepository.java#L249

If we like to go with 409 I suggest to open a JIRA ticket to fix the above contract/implementation 

the Item deletions are different 
https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java#L196
https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java#L200

as the client cannot "solve" the conflicts (if it is a templateItem or an inprogress submission it needs to use a different endpoint, there will be no a second call to the delete after "solving").

benbosman replied:
I've included it in this contract.
For the others a JIRA ticket will indeed be required

abollini commented on epersongroups.md:
````diff
@@ -0,0 +1,310 @@
+* GET /rest/api/eperson/groups/3b1de75d-5bf9-4ca3-bf4d-42bc8abd0d42/epersons?all=true
+
+
+```json
````
better avoid examples here as it will be difficult to keep all of them up-to-dated

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,310 @@
+
+### Remove an EPerson from a parent Group
+
+**DELETE /api/eperson/groups/<:groupuuid>/epersons**
```
same than for the groups, the DELETE should specify the eperson to remove from the association in the URL

benbosman replied:
This has been adjusted to be consistent with mapping collections

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,310 @@
+* 204: if the update succeeded 
+* 401 Forbidden - if you are not authenticated
+* 403 Unauthorized - if you are not logged in with sufficient permissions 
+* 422: if the specified eperson is not found
```
this is ambiguous. I guess that we mean that the specified eperson is not found in the system, it doesn't exist at all. It can be misunderstood as the specified eperson is not found in the group. In this later case, I guess that 204 should be returned with a no-op

benbosman replied:
This has been been updated

abollini commented on epersons.md:
```diff
@@ -2,10 +2,53 @@
 
 ## Main Endpoint
-**/api/eperson/epersons**  
+**GET /api/eperson/epersons**
```
yes, the current code use eperson as category as we have reserved "core" for the entities more specific of the DSpace domain
https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/model/EPersonRest.java#L25

No a strong opinion here, it was an early stage decision if we found a better way to structure such endpoints I'm absolutely open to discussion

benbosman commented on epersongroups.md:
```diff
@@ -0,0 +1,310 @@
+
+### Remove a Group from a parent Group
+
+**DELETE /api/eperson/groups/<:parentgroupuuid>/groups**
```
This has been adjusted to be consistent with mapping collections

cwilper commented:
Note that metadata-as-map has since been merged, so examples given in this PR should be updated to use that new structure, e.g. what I did here: https://github.com/DSpace/Rest7Contract/pull/53

benbosman commented on epersons.md:
```diff
@@ -74,3 +127,75 @@ the replace operation `[{ "op": "replace", "path": "/password", "value": "newpas
+* last name
+* email address
+
+## Linked entities
```
This was a copy-paste error.
It has been solved

benbosman commented on epersongroups.md:
```diff
@@ -0,0 +1,390 @@
+      "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc/groups"
+    },
+    "parentgroups": {
+      "href": "https://dspace7.4science.it/dspace-spring-rest/api/eperson/groups/617cf46b-535c-42d5-9d22-327ce2eff6dc/parentgroups"
```
Wouldn't `memberships` risk confusion about whether these are groups you're a member of vs members of this group?
Would `isMemberOf` be a good alternative

abollini replied:
I suggest to only include
- groups (that is also what is currently implemented)
- epersons

the membership should be retrieved with a search method

abollini replied:
please note that epersons would be problematic until we solve
https://github.com/DSpace/Rest7Contract/issues/2#issuecomment-528386269

as the eperson objects contains a reference to the group where the eperson is member

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,390 @@
+**GET /api/eperson/groups/<:uuid>/parentgroups**
+
+Retrieves direct parent groups of the given group.
+Requesting indirect parent groups is not supported
```
As noted above, I suggest to remove this association endpoint and add a search method

`/api/eperson/groups/search/hasMember?:uuid`

that will retrun the list of groups that contains the specified group by uuid

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,390 @@
+
+### Get Sub Groups in a single parent EPerson Group
+
+**GET /api/eperson/groups/<:uuid>/subgroups**
```
this should become /groups as in the current implementation

benbosman replied:
I actually used subgroups to make sure it's clear which groups it returns.
This ensures it's in line with https://github.com/DSpace/Rest7Contract/blob/master/communities.md#subcommunities

The old implementation (which didn't even have a REST contract) shouldn't be a reason to use harder to read URLs

tdonohue replied:
@benbosman : I'm OK with that approach. I was working under the assumption that we were attempting to align this contract with the current implementation. However, if you'd rather *change* the current implementation, I'm OK with renaming this subresource to `subgroups`. Thanks for clarifying your goal here.

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,390 @@
+* 401 Forbidden - if you are not authenticated
+* 403 Unauthorized - if you are not logged in with sufficient permissions 
+* 404 Not found - if the parent group doesn't exist
+* 409 Conflict - if adding the group would create a cyclic reference (ie. A -> B -> C, -> mean contains we cannot add A to C)
```
I'm not sure if we can use the 409 for anything different than versioning as this is the typical scenario where 409 is used. Maybe we should just use the 422 and think about adding later more details in the response body

tdonohue replied:
Yes, I think this could just be changed to a 422.  It looks (to me) like 409 is supposed to be for PUT requests that update a resource in an incorrect manner.

benbosman replied:
I've modified this to use 422

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,390 @@
+
+**DELETE /api/eperson/groups/<:parent_group_uuid>/subgroups**
+
+_Unsupported._ At this time, we do not support removing all subgroups in a single request. Please use `DELETE /api/core/groups/<:parent_group_uuid>/groups/<:sub_group_uuid>` to remove child groups one by one.
```
For consistency, I would suggest to don't include unsupported endpoint in the documentation as we haven't do that so far

abollini commented on epersons.md:
````diff
@@ -74,3 +127,58 @@ the replace operation `[{ "op": "replace", "path": "/password", "value": "newpas
+  "type": "eperson"
+}
+```
+
````
if you have time it would be useful to add the error code here explaining that the endpoint can be used only by administrators, otherwise we can improve that later

abollini commented on epersongroups.md:
````diff
@@ -0,0 +1,390 @@
+  }
+}
+```
+
````
it would be useful to add a sentence about the permanent attribute, i.e. it is a system required group that cannot be removed.

Moreover, we should also decide if the administrator is allowed or not to patch this information (or any metadata of permanent groups) and note that the DELETE of a group flagged as permanent should fail

benbosman replied:
I've included this description, it's easy to understand

benbosman commented:
I've updated this PR to prepare it for the beta 2 sprint implementation.
There are quite a few small fixes to the contract, and I've removed the parentgroups endpoint completely since it's not even implemented in the service, so most likely not relevant

tdonohue commented on epersongroups.md:
```diff
@@ -0,0 +1,307 @@
+
+### Add a Group to a parent Group
+
+**POST /api/eperson/groups/<:parentgroupuuid>/subgroups**
```
Again, this could just be `/groups/<:parentgroupuuid>/groups` like the current implementation.  We don't seem to use "subgroups" anywhere...so you could do a bulk replace throughout this PR

benbosman replied:
See https://github.com/DSpace/Rest7Contract/pull/41#discussion_r375177223

tdonohue commented on epersongroups.md:
```diff
@@ -0,0 +1,307 @@
+
+_Unsupported._ If you want detailed information about a single group, use the `/api/eperson/groups/<:uuid>` endpoint.
+
+**PUT /api/eperson/groups/<:parent_group_uuid>/subgroups**
```
This PUT section seems like it should be moved up alongside the `POST` that is similar.  It's a bit oddly placed under a header called "Remove a Group from a parent group".

benbosman replied:
I've moved this to make it more clear which methods are supported and which not for changing the subgroups

tdonohue commented on epersongroups.md:
```diff
@@ -0,0 +1,307 @@
+
+### Remove an EPerson from a parent Group
+
+**PUT /api/eperson/groups/<:groupuuid>/epersons**
```
Again, I think this PUT should be in the section above next to the POST

benbosman replied:
I've moved this to make it more clear which methods are supported and which not for changing the epeople

tdonohue commented on epersongroups.md:
```diff
@@ -0,0 +1,307 @@
+
+**GET /api/eperson/groups/<:parent_group_uuid>/subgroups/<:sub_group_uuid>**
+
+_Unsupported._ If you want detailed information about a single group, use the `/api/eperson/groups/<:uuid>` endpoint.
```
We can probably remove this entire GET section as it's unsupported.  I don't think it's necessary here.

benbosman replied:
I've removed this GET section

tdonohue commented on epersons.md:
```diff
@@ -74,3 +127,64 @@ the replace operation `[{ "op": "replace", "path": "/password", "value": "newpas
+This will return a pageable list of the groups this person is a direct member of
+
+> TODO: A solution to retrieve direct and indirect groups of an eperson is also required.
+> This would use GroupService.allMemberGroupsSet() and is used e.g. when viewing an EPerson as an admin
```
This TODO is a bit odd here (as it talks about code changes).  But, if we don't have a better place to put this, we can leave it here temporarily.

benbosman replied:
It doesn't require service or API code changes, but just states we'll need an additional rest endpoint to find all groups of an EPerson.
At this point in time I'd prefer not to include more endpoints in this PR because it's already a rather big PR for a contract.
The GroupService.allMemberGroupsSet can provide the functionality we need, but doesn't support paging

tdonohue replied:
That's fine. We'll just want to clean up this TODO statement at a later time then.

benbosman commented on epersongroups.md:
```diff
@@ -0,0 +1,307 @@
+* 201 Created - if the operation succeed
+* 401 Unauthorized - if you are not authenticated
+* 403 Forbidden - if you are not logged in with sufficient permissions
+* 422 Unprocessable Entity - if the name was omitted or already exists, if permanent was set to true
```
@tdonohue how I see this is:
`permanent` is a property on the group which is part of the JSON response. It's possible this is set to true during the POST (Angular should not be implemented to set it to true, but it can still be sent as such to REST from other applications). Since we should not allow creating permanent groups via REST, it should fail

tdonohue replied:
@benbosman : Oh, I understand what you mean now. You are correct, but I previously misunderstood the statement without the additional context you just added.  I'd suggest we simply reword this slightly and say: 

`422 Unprocessable Entity - if the name was omitted or already exists, *or* if permanent was set to true (permanent groups can only be created by the system).`

benbosman commented on epersongroups.md:
```diff
@@ -0,0 +1,307 @@
+          "permanent": true,
+          "type": "group",
+          "_links": {
+            "subgroups": {
```
See https://github.com/DSpace/Rest7Contract/pull/41#discussion_r375177223

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,303 @@
+Return codes:
+* 204: if the update succeeded 
+* 401 Forbidden - if you are not authenticated
+* 403 Unauthorized - if you are not logged in with sufficient permissions 
```
401 and 403 are inverted

benbosman replied:
This has been solved now

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,303 @@
+Return codes:
+* 204: if the update succeeded (including the case of no-op if the eperson was not a member)
+* 401 Forbidden - if you are not authenticated
+* 403 Unauthorized - if you are not logged in with sufficient permissions 
```
401 and 403 are inverted

benbosman replied:
This has been solved now

abollini commented on epersongroups.md:
```diff
@@ -0,0 +1,303 @@
+* 403 Unauthorized - if you are not logged in with sufficient permissions 
+* 404 Not found - if the parent group doesn't exist
+* 422 Unprocessable Entity - if the child group doesn't exist
+* 422 Unprocessable Entity - if the specified eperson doesn't exist
```
just a style note, we usually combine the different scenarios in a single sentence

benbosman replied:
This has been adjusted now

abollini commented on epersons.md:
```diff
@@ -74,3 +127,64 @@ the replace operation `[{ "op": "replace", "path": "/password", "value": "newpas
+
+This supports a basic search in the metadata.
+It will search in:
+* UUID (exact match)
```
if this is the already implemented behavior I'm fine to keep it, otherwise I would suggest to exclude UUID from the search as it is "by Metadata"

benbosman replied:
I did indeed base this on the current implementation to avoid having to create a different service implementation

tdonohue commented:
Merging, as this is at +2
